File: ZKSegment/ZKSegment.m
//
//  ZKSegment.m
//  ZKSegment
//
//  Created by WangWenzhuang on 15/10/20.
//  Copyright © 2015年 WangWenzhuang. All rights reserved.
//

#import "ZKSegment.h"

#define ZK_UIColorFromRGBAlpha(r, g, b, a) [UIColor colorWithRed:(r) / 255.0 green:(g) / 255.0 blue:(b) / 255.0 alpha:(a)]
#define ZK_ItemFontSize 14.0f
#define ZK_ItemMargin 20.0f
#define ZK_ItemPadding 8.0f
#define ZK_ScreenWidth [UIScreen mainScreen].bounds.size.width

#define ZK_Version @"1.0.2"

@interface ZKSegment ()
@property(nonatomic, strong) UIView *buttonStyle;
@property(nonatomic, assign) CGFloat maxWidth;
@property(nonatomic, strong) NSMutableArray *buttonList;
@property(nonatomic, strong) NSMutableArray *allItems;
@property(nonatomic, weak) UIButton *buttonSelected;
@property(nonatomic, assign) ZKSegmentStyle segmentStyle;
@property(nonatomic, assign) CGFloat buttonStyleY;
@property(nonatomic, assign) CGFloat buttonStyleHeight;
@property(nonatomic, assign) BOOL buttonStyleMasksToBounds;
@property(nonatomic, assign) CGFloat buttonStyleCornerRadius;
@end

@implementation ZKSegment

+ (ZKSegment *)zk_segmentWithFrame:(CGRect)frame style:(ZKSegmentStyle)style {
    return [[ZKSegment alloc] zk_initWithFrame:frame style:style];
}

#pragma 初始化
- (id)init {
    if (self = [super init]) {
        self.segmentStyle = ZKSegmentLineStyle;
        [self commonInit];
    }
    return self;
}

- (id)initWithCoder:(NSCoder *)coder {
    if ([super initWithCoder:coder]) {
        self.segmentStyle = ZKSegmentLineStyle;
        [self commonInit];
    }
    return self;
}

- (id)initWithFrame:(CGRect)frame {
    if ([super initWithFrame:frame]) {
        self.segmentStyle = ZKSegmentLineStyle;
        [self commonInit];
    }
    return self;
}

- (id)zk_initWithFrame:(CGRect)frame style:(ZKSegmentStyle)style {
    if ([super initWithFrame:frame]) {
        self.segmentStyle = *(&(style));
        [self commonInit];
    }
    return self;
}

- (void)commonInit {
    [self buttonStyleFromSegmentStyle];
    self.zk_itemDefaultColor = ZK_UIColorFromRGBAlpha(102.0f, 102.0f, 102.0f, 1.0f);
    self.itemStyleSelectedColor = ZK_UIColorFromRGBAlpha(202.0f, 51.0f, 54.0f, 1.0f);
    switch (self.segmentStyle) {
        case ZKSegmentLineStyle:
            self.zk_itemSelectedColor = ZK_UIColorFromRGBAlpha(202.0f, 51.0f, 54.0f, 1.0f);
            break;
        case ZKSegmentRectangleStyle:
            self.zk_itemSelectedColor = ZK_UIColorFromRGBAlpha(250.0f, 250.0f, 250.0f, 1.0f);
            break;
        case ZKSegmentTextStyle:
            self.zk_itemSelectedColor = ZK_UIColorFromRGBAlpha(202.0f, 51.0f, 54.0f, 1.0f);
            break;
    }
    self.segmentBackgroundColor = ZK_UIColorFromRGBAlpha(238.0f, 238.0f, 238.0f, 1.0f);
    self.maxWidth = ZK_ItemMargin;
    self.buttonList = [NSMutableArray array];
    self.allItems = [NSMutableArray array];
    self.bounces = NO;
    self.showsHorizontalScrollIndicator = NO;
    self.showsVerticalScrollIndicator = NO;
    self.buttonStyle = [[UIView alloc] initWithFrame:CGRectMake(ZK_ItemMargin, self.buttonStyleY, 0, self.buttonStyleHeight)];
    self.buttonStyle.backgroundColor = self.zk_itemStyleSelectedColor;
    self.buttonStyle.layer.masksToBounds = self.buttonStyleMasksToBounds;
    self.buttonStyle.layer.cornerRadius = self.buttonStyleCornerRadius;
    [self addSubview:self.buttonStyle];
}

- (void)setItemStyleSelectedColor:(UIColor *)itemStyleSelectedColor {
    _zk_itemStyleSelectedColor = itemStyleSelectedColor;
}

- (void)setSegmentBackgroundColor:(UIColor *)segmentBackgroundColor {
    _zk_backgroundColor = segmentBackgroundColor;
    self.backgroundColor = segmentBackgroundColor;
}

- (int)zk_selectedItemIndex {
    if (self.buttonSelected) {
        return [self indexOfItemsWithItem:self.buttonSelected.titleLabel.text];
    }
    return -1;
}

- (NSString *)zk_selectedItem {
    if (self.buttonSelected) {
        return self.buttonSelected.titleLabel.text;
    }
    return nil;
}

- (NSString *)zk_version {
    return ZK_Version;
}

#pragma 暴露给外部的
- (void)zk_setItems:(NSArray *)items {
    if (!items || items.count == 0) {
        return;
    }
    for (int i = 0; i < self.subviews.count; i++) {
        if (self.subviews[i] != self.buttonStyle) {
            [self.subviews[i--] removeFromSuperview];
        }
    }
    self.buttonStyle.hidden = NO;
    self.maxWidth = ZK_ItemMargin;
    [self.allItems removeAllObjects];
    [self.allItems addObjectsFromArray:items];
    self.buttonList = nil;
    self.buttonList = [[NSMutableArray alloc] init];
    self.buttonSelected = nil;
    if (self.allItems && self.allItems.count > 0) {
        self.backgroundColor = self.backgroundColor;
        for (int i = 0; i < self.allItems.count; i++) {
            [self createItem:self.allItems[i]];
        }
        self.contentSize = CGSizeMake(self.maxWidth, -4);
        [self fiexButtonWidth];
    }
}

- (NSArray *)zk_items {
    return self.allItems;
}

- (void)zk_itemClickByIndex:(NSInteger)index {
    if (index < 0) {
        return;
    }
    UIButton *item = (UIButton *)self.buttonList[index];
    [self itemClick:item];
}

- (void)zk_addItem:(NSString *)item {
    if (self.buttonList.count == 0) {
        self.buttonStyle.hidden = NO;
    }
    [self.allItems addObject:item];
    [self createItem:item];
    [self resetButtonsFrame];
    [self fiexButtonWidth];
}

- (void)zk_removeItemAtIndex:(NSInteger)index {
    if (self.allItems.count == 0 || index < 0 || index >= self.allItems.count) {
        return;
    }
    [self.allItems removeObjectAtIndex:index];
    UIButton *button = self.buttonList[index];
    [self.buttonList removeObjectAtIndex:index];
    for (int i = 0; i < self.subviews.count; i++) {
        if (self.subviews[i] == button) {
            [self.subviews[i] removeFromSuperview];
        }
    }
    if (button == self.buttonSelected && self.buttonList.count > 0) {
        NSInteger _index = index;
        if (self.buttonList.count >= index && index > 0) {
            _index = index - 1;
        }
        [self itemClick:self.buttonList[_index]];
    }
    if (self.buttonList.count == 0) {
        self.buttonSelected = nil;
        self.buttonStyle.hidden = YES;
    }
    [self resetButtonsFrame];
    [self fiexButtonWidth];
}

- (void)zk_removeItem:(NSString *)item {
    if (self.allItems.count == 0 || ![self.allItems containsObject:item]) {
        return;
    }
    for (NSInteger i = 0; i < self.allItems.count; i++) {
        if ([self.allItems[i] isEqualToString:item]) {
            [self zk_removeItemAtIndex:i];
            return;
        }
    }
}

#pragma 私有的
- (void)buttonStyleFromSegmentStyle {
    switch (self.segmentStyle) {
        case ZKSegmentLineStyle:
            self.buttonStyleY = self.frame.size.height - 2;
            self.buttonStyleHeight = 2.0f;
            break;
        case ZKSegmentRectangleStyle:
            self.buttonStyleY = ZK_ItemPadding;
            self.buttonStyleHeight = self.frame.size.height - ZK_ItemPadding * 2;
            self.buttonStyleCornerRadius = 6.0f;
            self.buttonStyleMasksToBounds = YES;
            break;
        case ZKSegmentTextStyle:
            self.buttonStyleY = 0;
            self.buttonStyleHeight = 0.0f;
            break;
    }
}

- (void)fiexButtonWidth {
    if (ZK_ScreenWidth - self.maxWidth > ZK_ItemMargin) {
        CGFloat bigButtonSumWidth = 0;
        int bigButtonCount = 0;
        self.maxWidth = ZK_ItemMargin;
        CGFloat width =
        (ZK_ScreenWidth - (self.buttonList.count + 1) * ZK_ItemMargin) /
        self.buttonList.count;
        for (int i = 0; i < self.buttonList.count; i++) {
            UIButton *button = self.buttonList[i];
            if (button.frame.size.width > width) {
                bigButtonCount++;
                bigButtonSumWidth += button.frame.size.width;
            }
        }
        width = (ZK_ScreenWidth - (self.buttonList.count + 1) * ZK_ItemMargin - bigButtonSumWidth) / (self.buttonList.count - bigButtonCount);
        for (int i = 0; i < self.buttonList.count; i++) {
            UIButton *button = self.buttonList[i];
            if (button.frame.size.width < width) {
                button.frame =
                CGRectMake(self.maxWidth, 0, width, self.frame.size.height);
                self.maxWidth += width + ZK_ItemMargin;
            } else {
                button.frame = CGRectMake(self.maxWidth, 0, button.frame.size.width, self.frame.size.height);
                self.maxWidth += button.frame.size.width + ZK_ItemMargin;
            }
            if (button == self.buttonSelected) {
                self.buttonStyle.frame = CGRectMake(button.frame.origin.x, self.buttonStyleY, button.frame.size.width, self.buttonStyleHeight);
            }
        }
        self.contentSize = CGSizeMake(self.maxWidth, -4);
    }
}

- (void)resetButtonsFrame {
    self.maxWidth = ZK_ItemMargin;
    
    for (int i = 0; i < self.allItems.count; i++) {
        CGFloat width = [self itemWidthFromSegmentStyle:self.allItems[i]];
        UIButton *button = self.buttonList[i];
        button.frame = CGRectMake(self.maxWidth, 0, width, self.frame.size.height);
        if (button == self.buttonSelected) {
            self.buttonStyle.frame = CGRectMake(self.maxWidth, self.buttonStyleY, width, self.buttonStyleHeight);
        }
        if (!self.buttonSelected) {
            [button setTitleColor:self.zk_itemSelectedColor forState:0];
            [self itemClick:button];
        }
        self.maxWidth += width + ZK_ItemMargin;
    }
    
    self.contentSize = CGSizeMake(self.maxWidth, -4);
}

- (void)createItem:(NSString *)item {
    CGFloat itemWidth = [self itemWidthFromSegmentStyle:item];
    UIButton *buttonItem = [[UIButton alloc] initWithFrame:CGRectMake(self.maxWidth, 0, itemWidth, self.frame.size.height)];
    buttonItem.titleLabel.font = [UIFont systemFontOfSize:ZK_ItemFontSize];
    [buttonItem setTitle:item forState:UIControlStateNormal];
    [buttonItem setTitleColor:self.zk_itemDefaultColor forState:UIControlStateNormal];
    [buttonItem addTarget:self action:@selector(itemClick:) forControlEvents:UIControlEventTouchUpInside];
    self.maxWidth += itemWidth + ZK_ItemMargin;
    [self.buttonList addObject:buttonItem];
    [self addSubview:buttonItem];
    if (!self.buttonSelected) {
        [buttonItem setTitleColor:self.zk_itemSelectedColor forState:0];
        self.buttonStyle.frame = CGRectMake(buttonItem.frame.origin.x, self.buttonStyleY, buttonItem.frame.size.width, self.buttonStyleHeight);
        [self itemClick:buttonItem];
    }
}

- (CGFloat)itemWidthFromSegmentStyle:(NSString *)item {
    CGFloat itemWidth = [self textWidthWithFontSize:ZK_ItemFontSize Text:item];
    CGFloat resultItemWidht;
    switch (self.segmentStyle) {
        case ZKSegmentLineStyle:
            resultItemWidht = itemWidth;
            break;
        case ZKSegmentRectangleStyle:
            resultItemWidht = itemWidth + ZK_ItemPadding * 2;
            break;
        case ZKSegmentTextStyle:
            resultItemWidht = itemWidth;
            break;
    }
    return resultItemWidht;
}

- (void)itemClick:(id)sender {
    UIButton *button = sender;
    if (self.buttonSelected != button) {
        [self.buttonSelected setTitleColor:self.zk_itemDefaultColor forState:0];
        [button setTitleColor:self.zk_itemSelectedColor forState:0];
        self.buttonSelected = button;
        if (self.zk_itemClickBlock) {
            int selectedIndex = [self indexOfItemsWithItem:button.titleLabel.text];
            NSString *selectedItem = button.titleLabel.text;
            self.zk_itemClickBlock(selectedItem, selectedIndex);
        }
        [UIView animateWithDuration:0.3
                              delay:0
                            options:UIViewAnimationOptionCurveEaseInOut
                         animations:^{
                             self.buttonStyle.frame = CGRectMake(button.frame.origin.x, self.buttonStyleY, button.frame.size.width, self.buttonStyleHeight);
                         }
                         completion:^(BOOL finished) {
                             [UIView animateWithDuration:
                              0.3 animations:^{
                                  CGFloat buttonX = button.frame.origin.x;
                                  CGFloat buttonWidth = button.frame.size.width;
                                  CGFloat scrollerWidth = self.contentSize.width;
                                  // 移动到中间
                                  if (scrollerWidth > ZK_ScreenWidth && // Scroller的宽度大于屏幕宽度
                                      buttonX > ZK_ScreenWidth / 2.0f - buttonWidth / 2.0f && //按钮的坐标大于屏幕中间位置
                                      scrollerWidth > buttonX + ZK_ScreenWidth / 2.0f + buttonWidth / 2.0f // Scroller的宽度大于按钮移动到中间坐标加上屏幕一半宽度加上按钮一半宽度
                                      ) {
                                      self.contentOffset = CGPointMake(button.frame.origin.x - ZK_ScreenWidth / 2.0f + button.frame.size.width / 2.0f, 0);
                                  } else if (buttonX < ZK_ScreenWidth / 2.0f - buttonWidth / 2.0f) { // 移动到开始
                                      self.contentOffset = CGPointMake(0, 0);
                                  } else if (scrollerWidth - buttonX < ZK_ScreenWidth / 2.0f + buttonWidth / 2.0f || // Scroller的宽度减去按钮的坐标小于屏幕的一半，移动到最后
                                             buttonX + buttonWidth + ZK_ItemMargin == scrollerWidth) {
                                      if (scrollerWidth > ZK_ScreenWidth) {
                                          self.contentOffset = CGPointMake(scrollerWidth - ZK_ScreenWidth, 0); // 移动到末尾
                                      }
                                  }
                              }];
                         }];
    }
}

- (int)indexOfItemsWithItem:(NSString *)item {
    for (int i = 0; i < self.allItems.count; i++) {
        if ([item isEqualToString:self.allItems[i]]) {
            return i;
        }
    }
    return -1;
}

- (CGFloat)textWidthWithFontSize:(CGFloat)fontSize Text:(NSString *)text {
    NSDictionary *attr = @{NSFontAttributeName : [UIFont systemFontOfSize:fontSize]};
    CGRect size = [text boundingRectWithSize:CGSizeMake(MAXFLOAT, self.frame.size.height)
                       options:NSStringDrawingUsesFontLeading | NSStringDrawingUsesLineFragmentOrigin
                    attributes:attr
                       context:nil];
    return size.size.width;
}

@end

File: ZKSegment/ZKSegment.swift
//
//  ZKSegment.swift
//  Demo
//
//  Created by 王文壮 on 2018/3/22.
//  Copyright © 2018年 王文壮. All rights reserved.
//

import UIKit

//MARK: 样式，默认：line
enum ZKSegmentStyle {
    case line
    case rectangle
    case text
    case dot
}

typealias Style = ZKSegmentStyle

public typealias ZKItemChange = ((_ index: Int, _ text: String) -> Void)

public extension ZKSegment {
    //MARK: 创建 .line 样式实例
    public static func segmentLine(frame: CGRect,
                                   itemColor: UIColor,
                                   itemSelectedColor: UIColor,
                                   itemFont: UIFont?,
                                   itemMargin: CGFloat?,
                                   items: [String] = [],
                                   change: ZKItemChange?) -> ZKSegment{
        return ZKSegment(
            frame: frame,
            style: .line,
            itemColor: itemColor,
            itemSelectedColor: itemSelectedColor,
            itemStyleSelectedColor: itemSelectedColor,
            itemFont: itemFont,
            itemMargin: itemMargin,
            items: items,
            change: change)
    }
    //MARK: 创建 .rectangle 样式实例
    public static func segmentRectangle(frame: CGRect,
                                        itemColor: UIColor,
                                        itemSelectedColor: UIColor,
                                        itemStyleSelectedColor: UIColor,
                                        itemFont: UIFont?,
                                        itemMargin: CGFloat?,
                                        items: [String] = [],
                                        change: ZKItemChange?) -> ZKSegment{
        return ZKSegment(
            frame: frame,
            style: .rectangle,
            itemColor: itemColor,
            itemSelectedColor: itemSelectedColor,
            itemStyleSelectedColor: itemStyleSelectedColor,
            itemFont: itemFont,
            itemMargin: itemMargin,
            items: items,
            change: change)
    }
    //MARK: 创建 .text 样式实例
    public static func segmentText(frame: CGRect,
                                   itemColor: UIColor,
                                   itemSelectedColor: UIColor,
                                   itemFont: UIFont?,
                                   itemMargin: CGFloat?,
                                   items: [String] = [],
                                   change: ZKItemChange?) -> ZKSegment{
        return ZKSegment(
            frame: frame,
            style: .text,
            itemColor: itemColor,
            itemSelectedColor: itemSelectedColor,
            itemStyleSelectedColor: .clear,
            itemFont: itemFont,
            itemMargin: itemMargin,
            items: items,
            change: change)
    }
    //MARK: 创建 .dot 样式实例
    public static func segmentDot(frame: CGRect,
                                   itemColor: UIColor,
                                   itemSelectedColor: UIColor,
                                   itemFont: UIFont?,
                                   itemMargin: CGFloat?,
                                   items: [String] = [],
                                   change: ZKItemChange?) -> ZKSegment{
        return ZKSegment(
            frame: frame,
            style: .dot,
            itemColor: itemColor,
            itemSelectedColor: itemSelectedColor,
            itemStyleSelectedColor: itemSelectedColor,
            itemFont: itemFont,
            itemMargin: itemMargin,
            items: items,
            change: change)
    }
}

public class ZKSegment: UIScrollView {
    //MARK: 获取当前选中项索引
    public var selectedIndex: Int? {
        get {
            if self.buttonSelected != nil {
                if let index = self.items.index(where: { $0 == self.buttonSelected?.titleLabel?.text }) {
                    return index
                }
            }
            return nil
        }
    }
    //MARK: 获取当前选中项
    public var selectedItem: String? {
        get {
            if self.buttonSelected != nil {
                return self.buttonSelected?.titleLabel?.text
            }
            return nil
        }
    }
    
    //MARK: 项更改事件
    private var itemChange: ZKItemChange?
    //MARK: 每一项颜色
    private var itemColor: UIColor!
    //MARK: 选中项颜色
    private var itemSelectedColor: UIColor!
    //MARK: 选中项样式颜色
    private var itemStyleSelectedColor: UIColor!
    //MARK: 每一项字体
    private var itemFont: UIFont!
    //MARK: 每一项间距
    private var itemMargin: CGFloat = 20
    //MAKR: 样式
    private var style: Style = .line
    
    private var buttons: [UIButton] = []
    private var buttonSelected: UIButton?
    
    private var itemStyle: UIView!
    private var itemStyleY: CGFloat = 0
    private var itemStyleHeight: CGFloat = 0
    
    //MARK: .rectangle 样式圆角属性
    private let itemRectangleStyleCornerRadius: CGFloat = 6
    //MARK: .rectangle 样式每项内间距
    private let itemRectangleStylePadding: CGFloat = 8
    
    private var contentX: CGFloat = 0
    
    private var items: [String] = []
    
    private init(frame: CGRect,
                 style: Style,
                 itemColor: UIColor,
                 itemSelectedColor: UIColor,
                 itemStyleSelectedColor: UIColor,
                 itemFont: UIFont?,
                 itemMargin: CGFloat?,
                 items: [String],
                 change: ZKItemChange?) {
        
        super.init(frame: frame)
        self.style = style
        self.itemColor = itemColor
        self.itemSelectedColor = itemSelectedColor
        self.itemStyleSelectedColor = itemStyleSelectedColor
        self.itemFont = itemFont ?? UIFont.systemFont(ofSize: 14)
        self.itemMargin = itemMargin ?? self.itemMargin
        self.itemChange = change
        
        self.showsHorizontalScrollIndicator = false
        self.showsVerticalScrollIndicator = false
        
        self.itemStyle = UIView()
        self.itemStyle.backgroundColor = self.itemStyleSelectedColor
        
        switch style {
        case .line:
            self.itemStyleY = self.height - 2
            self.itemStyleHeight = 2
        case .rectangle:
            self.itemStyleY = self.itemRectangleStylePadding
            self.itemStyleHeight = self.height - self.itemStyleY * 2
            self.itemStyle.layer.cornerRadius = self.itemRectangleStyleCornerRadius
            self.itemStyle.layer.masksToBounds = true
        case .text:
            self.itemStyleY = 0
            self.itemStyleHeight = 0
        case .dot:
            self.itemStyleY = self.height - 10
            self.itemStyleHeight = 4
            self.itemStyle.layer.cornerRadius = 2
            self.itemStyle.layer.masksToBounds = true
        }
        self.reload(items)
    }
    
    required public init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}

public extension ZKSegment {
    //MARK: 重新加载
    public func reload(_ items: [String]) {
        self.items = items
        self.buttons.removeAll()
        self.subviews.forEach({ $0.removeFromSuperview() })
        self.itemStyle.isHidden = false
        self.contentX = self.itemMargin
        self.addSubview(self.itemStyle)
        buttonSelected = nil
        self.items.forEach({ self.createItem($0) })
        self.contentSize = CGSize(width: self.contentX, height: -4)
        self.fiexItems()
    }
    //MARK: 根据索引选中
    public func select(_ index: Int) {
        if index >= 0 {
            self.itemClick(button: self.buttons[index])
        }
    }
    //MARK: 添加一项
    public func add(_ item: String, isSelected: Bool = false) {
        if self.itemStyle.isHidden {
            self.itemStyle.isHidden = false
        }
        self.items.append(item)
        self.createItem(item)
        self.resetitemsFrame()
        self.fiexItems()
    }
    //MARK: 移除一项
    public func remove(_ index: Int) {
        if index < 0 || index > self.items.count - 1 {
            print("ZKSegment ->>>>>> error: remove 方法索引不对")
            return
        }
        self.items.remove(at: index)
        let button = self.buttons[index]
        self.buttons.remove(at: index)
        button.removeFromSuperview()
        if button == self.buttonSelected && self.buttons.count > 0 {
            var itemIndex = index
            if self.buttons.count >= index && index > 0 {
                itemIndex = index - 1
            }
            self.itemClick(button: self.buttons[itemIndex])
        }
        if self.buttons.count == 0 {
            self.buttonSelected = nil
            self.itemStyle.isHidden = true
        }
        self.resetitemsFrame()
        self.fiexItems()
    }
}

extension ZKSegment {
    private func itemStyleFrame(_ x: CGFloat, _ width: CGFloat) -> CGRect {
        var styleX = x
        var styleWidth = width
        if self.style == .dot {
            styleWidth = 12
            styleX = x + (width - styleWidth) / 2
        }
        return CGRect(
            x: styleX,
            y: self.itemStyleY,
            width: styleWidth,
            height: self.itemStyleHeight
        )
    }
    
    private func createItem(_ item: String) {
        let itemWidth = self.itemWidth(item)
        let button = UIButton(frame: CGRect(x: self.contentX, y: 0, width: itemWidth, height: self.height))
        button.titleLabel?.font = self.itemFont
        button.setTitle(item, for: .normal)
        button.setTitleColor(self.itemColor, for: .normal)
        button.addTarget(self, action: #selector(ZKSegment.itemClick(button:)), for: .touchUpInside)
        self.contentX += (itemWidth + self.itemMargin)
        self.buttons.append(button)
        self.addSubview(button)
        if self.buttonSelected == nil {
            button.setTitleColor(self.itemSelectedColor, for: .normal)
            self.itemStyle.frame = self.itemStyleFrame(button.x, button.width)
            self.itemClick(button: button)
        }
    }
    
    @objc private func itemClick(button: UIButton) {
        if self.buttonSelected != button {
            if self.buttonSelected != nil {
                self.buttonSelected?.setTitleColor(self.itemColor, for: .normal)
            }
            button.setTitleColor(self.itemSelectedColor, for: .normal)
            self.buttonSelected = button
            if let change = self.itemChange {
                change(self.selectedIndex ?? -1, self.selectedItem ?? "")
            }
            self.animateChage(button)
        }
    }
    
    private func animateChage(_ button: UIButton) {
        UIView.animate(
            withDuration: 0.3,
            delay: 0,
            options: UIViewAnimationOptions.curveEaseOut,
            animations: {
                self.itemStyle.frame = self.itemStyleFrame(button.x, button.width)
        }, completion: { finished in
            UIView.animate(withDuration: 0.3, animations: {
                //MARK: 移动到中间
                if self.contentSize.width > self.width &&                                               // 内容宽度大于控件宽度
                    button.x > self.width / 2 - button.width / 2 &&                                     // 按钮的坐标大于屏幕中间位置
                    self.contentSize.width > button.x + self.width / 2 + button.width / 2 {             //内容的宽度大于按钮移动到中间坐标加上屏幕一半宽度加上按钮一半宽度
                    self.contentOffset = CGPoint(
                        x: button.x - self.width / 2 + button.width / 2,
                        y: 0)
                } else if button.x < self.width / 2 - button.width / 2 {                                // 移动到开始
                    self.contentOffset = CGPoint(x: 0, y: 0)
                } else if self.contentSize.width - button.x < self.width / 2 + button.width / 2 ||      // 内容宽度减去按钮的坐标小于屏幕的一半，移动到最后
                    button.x + button.width + self.itemMargin == self.contentSize.width {
                    if self.contentSize.width > self.width {
                        self.contentOffset = CGPoint(x: self.contentSize.width - self.width, y: 0)      // 移动到末尾
                    }
                }
            })
        })
    }
    //MARK: 根据样式获取项宽度
    private func itemWidth(_ item: String) -> CGFloat {
        let itemWidth = item.size(self.itemFont).width
        switch self.style {
        case .line,
             .text,
             .dot:
            return itemWidth
        case .rectangle:
            return itemWidth + self.itemMargin * 2
        }
    }
    //MARK: 如果内容宽度小于控件宽度，居中显示
    private func fiexItems() {
        if (self.width - self.contentX > self.itemMargin) {
            var bigItemSumWidth: CGFloat = 0
            var bigItemCount: CGFloat = 0
            self.contentX = self.itemMargin
            // 计算平均每项宽度
            var itemWidth = (self.width - (CGFloat(self.buttons.count) + 1) * self.itemMargin) / CGFloat(self.buttons.count)
            // 检查是否有超过平均宽度的项
            self.buttons.forEach({ button in
                if button.width > itemWidth {
                    bigItemCount += 1
                    bigItemSumWidth += button.width
                }
            })
            // 减去超过平均宽度项的宽度总和，重新计算剩余项的宽度
            itemWidth = (self.width - (CGFloat(self.buttons.count) + 1) * self.itemMargin - bigItemSumWidth) / (CGFloat(self.buttons.count) - bigItemCount)
            // 重新布局
            self.buttons.forEach({ button in
                // 如果小于平均宽度，设置为平均宽度
                if button.width < itemWidth {
                    button.frame = CGRect(x: self.contentX, y: 0, width: itemWidth, height: self.height)
                    self.contentX += (itemWidth + self.itemMargin)
                } else {
                    button.frame = CGRect(x: self.contentX, y: 0, width: button.width, height: self.height);
                    self.contentX += (button.width + self.itemMargin)
                }
                if button == self.buttonSelected {
                    self.itemStyle.frame = self.itemStyleFrame(button.x, button.width)
                }
            })
            self.contentSize = CGSize(width: self.contentX, height: -4)
        }
    }
    //MARK: 重置所有项 frame，如果调用了 fiexItems 可能按钮被拉伸了宽度，
    //      所以如果变更了项，需要重置一下所有项的 width
    private func resetitemsFrame() {
        self.contentX = self.itemMargin
        
        for index in 0...self.items.count - 1 {
            let itemWidth = self.itemWidth(self.items[index])
            let button = self.buttons[index]
            button.frame = CGRect(x: self.contentX, y: 0, width: itemWidth, height: self.height)
            if button == self.buttonSelected {
                self.itemStyle.frame = self.itemStyleFrame(self.contentX, itemWidth)
            }
            if self.buttonSelected == nil {
                button.setTitleColor(self.itemSelectedColor, for: .normal)
                self.itemClick(button: button)
            }
            self.contentX += (itemWidth + self.itemMargin)
        }
        self.contentSize = CGSize(width: self.contentX, height: -4)
    }
}

extension String {
    func size(_ font: UIFont) -> CGSize {
        let attribute = [ NSAttributedStringKey.font: font ]
        let conten = NSString(string: self)
        return conten.boundingRect(
            with: CGSize(width: CGFloat(MAXFLOAT), height: CGFloat(MAXFLOAT)),
            options: .usesLineFragmentOrigin,
            attributes: attribute,
            context: nil
            ).size
    }
}

extension UIView {
    var x: CGFloat {
        get {
            return self.frame.origin.x
        }
    }
    
    var y: CGFloat {
        get {
            return self.frame.origin.y
        }
    }
    
    var width: CGFloat {
        get {
            return self.frame.size.width
        }
    }
    
    var height: CGFloat {
        get {
            return self.frame.size.height
        }
    }
}
